### PR TITLE
feat(planner): support union/union all 

### LINKED ISSUE #7034: Feature: support union/union all for databend
**Summary**
https://www.postgresql.org/docs/current/queries-union.html (We have supported `intersect` and `except` set operator.)

`Union` also is used in recursive CTE.

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support union distinct and union all

Fixes #7034

### CODE REVIEW

File: src/query/service/src/sql/executor/pipeline_builder.rs [hunk 1]
```diff
@@ -472,4 +481,43 @@ impl PipelineBuilder {
...
+        self.pipelines.push(build_res.main_pipeline);
+        self.pipelines
+            .extend(build_res.sources_pipelines.into_iter());
+        Ok(rx)
+    }
+
+    pub fn build_union(&mut self, union: &Union) -> Result<()> {
+        self.build_pipeline(&union.left)?;
+        let union_receiver = self.expand_union(&union.right)?;
+        self.main_pipeline.resize(1)?;
```
Comment (zhang2014) [inline src/query/service/src/sql/executor/pipeline_builder.rs:509]: Not need this if we use async_channel::bounded(2)?

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    pub fn create(sender: Sender<Option<DataBlock>>, input: Arc<InputPort>) -> ProcessorPtr {
+        Sinker::create(input, UnionReceiveSink {
+            input_blocks: vec![],
+            sender,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for UnionReceiveSink {
```
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:42]:
```suggestion
impl AsyncSink for UnionReceiveSink {

}
```

File: src/query/service/src/sql/executor/pipeline_builder.rs (hunk 1 above)
Comment (xudong963) [reply src/query/service/src/sql/executor/pipeline_builder.rs]: Yes, fixed

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncSink for UnionReceiveSink {
+    const NAME: &'static str = "UnionReceiveSink";
+
+    async fn on_finish(&mut self) -> Result<()> {
+        let send_blocks = if self.input_blocks.is_empty() {
```
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:48]: Move this to `consume`?

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        if let Err(_) = self.sender.send(send_blocks).await {
+            return Err(ErrorCode::UnexpectedError("UnionReceiveSink sender failed"));
+        };
+        self.sender.close();
+        Ok(())
+    }
+
+    #[unboxed_simple]
+    async fn consume(&mut self, data_block: DataBlock) -> Result<()> {
+        self.input_blocks.push(data_block);
```
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:62]:
```suggestion
        self.sender.send(data_block);
```

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs (hunk 3 above)
Comment (xudong963) [reply src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:48]: fixed

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_pipeline_core::processors::port::InputPort;
+use common_pipeline_core::processors::processor::ProcessorPtr;
+
+use crate::processors::sinks::AsyncSink;
+use crate::processors::sinks::AsyncSinker;
+
+pub struct UnionReceiveSink {
+    sender: Sender<DataBlock>,
```
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:30]:
```suggestion
    sender: Option<Sender<DataBlock>>,
```

File: src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        AsyncSinker::create(input, UnionReceiveSink { sender })
+    }
+}
+
+#[async_trait]
+impl AsyncSink for UnionReceiveSink {
+    const NAME: &'static str = "UnionReceiveSink";
+
+    async fn on_finish(&mut self) -> Result<()> {
+        self.sender.close();
```
Comment (zhang2014) [inline src/common/pipeline/sinks/src/processors/sinks/union_receive_sink.rs:44]:
```suggestion
        self.sender.take();
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+        if let Some(output_data) = self.output_data.take() {
+            self.output.push_data(Ok(output_data));
+            return Ok(Event::NeedConsume);
+        }
+
+        if self.input_data.is_some() || self.receiver_result.is_some() {
+            return Ok(Event::Sync);
+        }
+
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:94]:
```suggestion
if let Ok(receiver_result) = self.receiver.try_recv() {
    self.receiver_result = receiver_result;
    return Ok(Event::Sync);
}
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    fn name(&self) -> &'static str {
+        "TransformMergeBlock"
+    }
+
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn event(&mut self) -> Result<Event> {
+        if !self.initialized {
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:72]: remove this branch.

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            self.output.push_data(Ok(output_data));
+            return Ok(Event::NeedConsume);
+        }
+
+        if self.input_data.is_some() || self.receiver_result.is_some() {
+            return Ok(Event::Sync);
+        }
+
+        if self.input.is_finished() {
+            self.output.finish();
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:96]:
```suggestion
            if !self.finished {
                return Ok(Event::Async);
            }

            self.output.finish();
```

File: src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        } else if let Some(receiver_result) = self.receiver_result.take() {
+            let data_block =
+                DataBlock::create(self.schema.clone(), receiver_result.columns().to_vec());
+            self.output_data = Some(data_block);
+        }
+
+        Ok(())
+    }
+
+    async fn async_process(&mut self) -> Result<()> {
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/transform_merge_block.rs:127]:
```suggestion
    async fn async_process(&mut self) -> Result<()> {
        if !self.finished   {
            if let Ok(receiver_result) = self.receiver.recv().await {
                    self.receiver_result = Some(receiver_result);
                    return Ok(());
            }

            self.finished = true;
        }

        Ok(())
    }
```

**Comment by xudong963** [comment]
@zhang2014 I think your comments are similar to the change https://github.com/datafuselabs/databend/pull/7160/commits/31b090fb850259ba684579a7459d8af027a60125 ?

File: src/query/service/src/sql/executor/pipeline_builder.rs [hunk 11]
```diff
@@ -472,4 +480,52 @@ impl PipelineBuilder {
...
+        Ok(rx)
+    }
+
+    pub fn build_union(&mut self, union: &Union) -> Result<()> {
+        self.build_pipeline(&union.left)?;
+        let union_receiver = self.expand_union(&union.right)?;
+        let mut inputs_port = Vec::with_capacity(self.main_pipeline.output_len());
+        let mut outputs_port = Vec::with_capacity(self.main_pipeline.output_len());
+        let mut processors = Vec::with_capacity(self.main_pipeline.output_len());
+        for _ in 0..self.main_pipeline.output_len() {
```
Comment (zhang2014) [inline src/query/service/src/sql/executor/pipeline_builder.rs:511]:
```suggestion
        self.main_pipeline.add_transform
```

**Comment by zhang2014** [comment]
> @zhang2014 I think your comments are similar to the change [31b090f](https://github.com/datafuselabs/databend/commit/31b090fb850259ba684579a7459d8af027a60125) ? Make receiver receive all messages

We should not wait for all data to be ready before merge `union` left and right

**Comment by xudong963** [comment]
> We should not wait for all data to be ready before merge `union` left and right

I got you, thanks for your detailed explanation, learn much!

**Review by zhang2014 — APPROVED** [review]

File: src/query/service/src/sql/executor/physical_plan.rs [hunk 12]
```diff
@@ -281,6 +281,19 @@ impl ExchangeSink {
     }
 }
 
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct Union {
```
Comment (leiysky) [inline src/query/service/src/sql/executor/physical_plan.rs:285]: Better rename it to `UnionAll`.

Comment (xudong963) [reply src/query/service/src/sql/executor/physical_plan.rs]: done

**Review by sundy-li — APPROVED** [review]

**Comment by xudong963** [comment]
@soyeric128 Please add union/union all to https://databend.rs/doc/reference/sql/query-operators/operators-set when you have time

**Comment by soyeric128** [comment]
> @soyeric128 Please add union/union all to https://databend.rs/doc/reference/sql/query-operators/operators-set when you have time

Ok

**Comment by soyeric128** [comment]
Please review https://github.com/datafuselabs/databend/pull/7281

**Pull request merged**